### PR TITLE
refactor!: migration to use attrs on the majority of models

### PR DESCRIPTION
**Pull request by Catalyst4222**
## About

This PR converts the DictSerializationMixin to use attrs instead of slots
Some benefits:
- Nested type conversion
- No massive `__init__`s
- The ability to rename data passed from discord into different attributes
- Stricter object creation (e.x. requiring HTTPClient for classes that make use of it)

This is very breaking

## Checklist

- [x] I've ran `pre-commit` to format and lint the change(s) made.
- [x] I've checked to make sure the change(s) work on `3.8.6` and higher.
- [ ] This fixes/solves an [Issue](https://github.com/goverfl0w/discord-interactions/issues).
  - (If existent):
- [x] I've made this pull request for/as: (check all that apply)
  - [ ] Documentation
  - [x] Breaking change
  - [x] New feature/enhancement
  - [x] Bugfix

### CODE REVIEW

**Comment by sillyElvy** [comment]
Is there a way `client` could stay underscored? That's very big breaking change

**Comment by sillyElvy** [comment]
Also why is this a feat and not a refactor?

**Comment by Catalyst4222** [comment]
> Is there a way client could stay underscored? That's very big breaking change
It would be possible, but attrs strips the leading underscore by default, and I've already changed most of the things to use client. I can figure out a way to keep the underscore if needed, but it would have to wait a bit
> Also why is this a feat and not a refactor?
Because I'm used to making only feat/fix 😅, changing that now

**Comment by sillyElvy** [comment]
> I can figure out a way to keep the underscore if needed, but it would have to wait a bit

I really don't want to let this PR through until that, a lot of users would have to change their code for this change and that's kinda bad

**Comment by Catalyst4222** [comment]
I said that I could do it, I just need time to do so

**Comment by sillyElvy** [comment]
I just wanted to clarify *why* I have a problem with it 😅

Alright, thanks

**Comment by i0bs** [comment]
Creating a property method to alias client as underscored may be the best approach for now

**Comment by AstreaTSS** [comment]
Napkining real quick - a way of allowing `_client` to still be a thing is by making a subclass of the mixin that has it. Something like so:

```python
@attrs.define(...) # same as the dict serializer
class ClientSerializerMixin(DictSerializerMixin):
    _client: "HTTPClient" = attrs.field(init=False, repr=False)

    def __init__(self, kwargs_dict: dict = None, **other_kwargs):
        kwargs = kwargs_dict or other_kwargs
        self._client = kwargs.get("_client")
        super().__init__(**kwargs)
```

File: interactions/api/models/misc.py
Comment (AstreaTSS) [inline interactions/api/models/misc.py]: `extras` should be underscored *just* in case Discord adds an attribute called just that.

Comment (Catalyst4222) [reply interactions/api/models/misc.py]: Changed in [7e1a139](https://github.com/interactions-py/library/pull/795/commits/7e1a13994ce64c992cc6ea550281bca170f3e6f0)

**Comment by AstreaTSS** [comment]
Heads up: I'll be reviewing this in small chunks due to the *huge* amount of code changes and will likely not do a super deep analysis into every piece of code.

File: interactions/api/gateway/client.py
Comment (AstreaTSS) [inline interactions/api/gateway/client.py:295]: Shouldn't this (and pretty much the entire file here) be using `_client`?

File: interactions/api/models/message.py
Comment (AstreaTSS) [inline interactions/api/models/message.py]: I *think* this repr can removed - `attrs` can just use it's usual repr. Using the content isn't *really* great for representing messages when many people won't even have the guild message content intent on.

I know, this kind of sounds like a thing for a separate PR, and it honestly might be, but this was likely here in the first place because there was no auto-generating repr. There is one now.

**Review by AstreaTSS — COMMENTED** [review]
Note: my web browser is lagging due to how big this PR is 😅 - I wanted to review a bit more, but I'll have to stop things here for now.

File: interactions/api/gateway/client.py
Comment (Catalyst4222) [reply interactions/api/gateway/client.py:295]: I believe `.client` is correct there, since the Context classes don't have it underscored

Comment (sillyElvy) [reply interactions/api/gateway/client.py:295]: I think it's for an option value here, not for context itself?

Comment (Catalyst4222) [reply interactions/api/gateway/client.py:295]: I've tested with options, and they work
That line isn't even touched in this pr

**Review by AstreaTSS — APPROVED** [review]
Looks good!

Heads up though: I don't really have anything to test the extensive amount of changes this PR does - I could only do some basic tests here and there. If anyone else sees this, more testing would be *very* appreciated.

**Review by sillyElvy — CHANGES_REQUESTED** [review]
I won't approve this until you are core dev /shrug

**Comment by Damego** [comment]
> I won't approve this until you are core dev /shrug

Now you should approve this pr

**Review by sillyElvy — CHANGES_REQUESTED** [review]
You have merge conflicts

**Comment by Catalyst4222** [comment]
This pr should be ready, all merge conflicts have been fixed and I have tested as much as I can remember

**Review by sillyElvy — APPROVED** [review]

**Pull request merged**
